File: frontend/src/app/spatial.models.ts
export interface Vector3Dto {
  x: number;
  y: number;
  z: number;
}

export interface QuaternionDto {
  x: number;
  y: number;
  z: number;
  w: number;
}

export interface Transform3DDto {
  translation: Vector3Dto;
  rotation: QuaternionDto;
}

export interface Bounds3DDto {
  width: number;
  height: number;
  depth: number;
}

export interface SpaceDto {
  id: string;
  name: string;
  worldTransform: Transform3DDto;
  createdAt: string;
  updatedAt: string;
}

export interface FurnitureInstanceDto {
  id: string;
  spaceId: string;
  name: string;
  category: string;
  spaceTransform: Transform3DDto;
  bounds: Bounds3DDto;
  recognitionMode: RecognitionMode;
  confidence: number;
  visualDescriptor: string | null;
  createdAt: string;
  updatedAt: string;
}

export type RecognitionMode =
  | 'MANUAL_BOUNDING_BOX'
  | 'VISUAL_RELOCALIZATION'
  | 'MARKER_ASSISTED'
  | 'AUTOMATIC';

export interface CreateSpaceRequest {
  name: string;
  worldTransform: Transform3DDto;
}

export interface RegisterFurnitureRequest {
  name: string;
  category: string;
  spaceTransform: Transform3DDto;
  bounds: Bounds3DDto;
  recognitionMode: RecognitionMode;
  confidence: number;
  visualDescriptor: string | null;
}

export interface FurnitureEditorState {
  spaceTransform: Transform3DDto;
  bounds: Bounds3DDto;
  yawDegrees: number;
}

export const IDENTITY_TRANSFORM: Transform3DDto = {
  translation: { x: 0, y: 0, z: 0 },
  rotation: { x: 0, y: 0, z: 0, w: 1 }
};

File: backend/src/main/java/io/arhome/spatial/domain/SpatialModels.java
package io.arhome.spatial.domain;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

import io.arhome.spatial.domain.SpatialTypes.Bounds3D;
import io.arhome.spatial.domain.SpatialTypes.Transform3D;

public final class SpatialModels {

    private SpatialModels() {
    }

    public enum RecognitionMode {
        MANUAL_BOUNDING_BOX,
        VISUAL_RELOCALIZATION,
        MARKER_ASSISTED,
        AUTOMATIC
    }

    public enum PositionPrecision {
        APPROXIMATE,
        ZONE,
        CENTIMETRIC
    }

    public enum PlacementMethod {
        MANUAL,
        AR_TAP,
        VISUAL_ESTIMATE,
        IMPORTED
    }

    public record Space(
            UUID id,
            String name,
            Transform3D worldTransform,
            Instant createdAt,
            Instant updatedAt) {

        public Space {
            Objects.requireNonNull(id, "id");
            requireText(name, "name");
            Objects.requireNonNull(worldTransform, "worldTransform");
            Objects.requireNonNull(createdAt, "createdAt");
            Objects.requireNonNull(updatedAt, "updatedAt");
        }
    }

    public record FurnitureInstance(
            UUID id,
            UUID spaceId,
            String name,
            String category,
            Transform3D spaceTransform,
            Bounds3D bounds,
            RecognitionMode recognitionMode,
            double confidence,
            String visualDescriptor,
            Instant createdAt,
            Instant updatedAt) {

        public FurnitureInstance {
            Objects.requireNonNull(id, "id");
            Objects.requireNonNull(spaceId, "spaceId");
            requireText(name, "name");
            requireText(category, "category");
            Objects.requireNonNull(spaceTransform, "spaceTransform");
            Objects.requireNonNull(bounds, "bounds");
            Objects.requireNonNull(recognitionMode, "recognitionMode");
            requireConfidence(confidence);
            Objects.requireNonNull(createdAt, "createdAt");
            Objects.requireNonNull(updatedAt, "updatedAt");
        }
    }

    public record Compartment(
            UUID id,
            UUID furnitureId,
            UUID parentCompartmentId,
            String name,
            String type,
            Transform3D parentTransform,
            Bounds3D bounds,
            Instant createdAt,
            Instant updatedAt) {

        public Compartment {
            Objects.requireNonNull(id, "id");
            Objects.requireNonNull(furnitureId, "furnitureId");
            requireText(name, "name");
            requireText(type, "type");
            Objects.requireNonNull(parentTransform, "parentTransform");
            Objects.requireNonNull(bounds, "bounds");
            Objects.requireNonNull(createdAt, "createdAt");
            Objects.requireNonNull(updatedAt, "updatedAt");
        }
    }

    public record ItemPlacement(
            String itemId,
            UUID furnitureId,
            UUID compartmentId,
            Transform3D localTransform,
            PositionPrecision precision,
            PlacementMethod method,
            double confidence,
            Instant createdAt,
            Instant updatedAt) {

        public ItemPlacement {
            requireText(itemId, "itemId");
            Objects.requireNonNull(furnitureId, "furnitureId");
            Objects.requireNonNull(localTransform, "localTransform");
            Objects.requireNonNull(precision, "precision");
            Objects.requireNonNull(method, "method");
            requireConfidence(confidence);
            Objects.requireNonNull(createdAt, "createdAt");
            Objects.requireNonNull(updatedAt, "updatedAt");
        }
    }

    public record ResolvedItemPose(
            String itemId,
            UUID spaceId,
            UUID furnitureId,
            UUID compartmentId,
            Transform3D worldTransform,
            PositionPrecision precision,
            double confidence) {
    }

    public record FurnitureCandidate(
            UUID furnitureId,
            String name,
            String category,
            double score,
            double positionScore,
            double geometryScore,
            double visualScore,
            double categoryScore) {
    }

    private static void requireText(String value, String field) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(field + " must not be blank");
        }
    }

    private static void requireConfidence(double confidence) {
        if (confidence < 0 || confidence > 1) {
            throw new IllegalArgumentException("confidence must be between 0 and 1");
        }
    }
}
